esphome/components/espnow/espnow.cpp

```diff
@@ -0,0 +1,549 @@
+#include "esphome/core/application.h"
+#include "esphome/core/version.h"
+#include "esphome/core/log.h"
+
+#if defined(USE_ESP32)
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(5, 0, 1)
+
+#include "espnow.h"
+
+#include <cstring>
+
+#include "esp_mac.h"
+#include "esp_random.h"
+
+#include "esp_wifi.h"
+#include "esp_now.h"
+#include "esp_event.h"
+
+#ifdef USE_WIFI
+#include "esphome/components/wifi/wifi_component.h"
+#endif
+
+#include <memory>
+
+namespace esphome {
+namespace espnow {
+
+static const char *const TAG = "espnow";
+const char *const ESPNowTAG::TAG = "espnow";
+
+static const esp_err_t ESP_ERR_ESPNOW_CMP_BASE = (ESP_ERR_ESPNOW_BASE + 20);
+static const esp_err_t ESP_ERR_ESPNOW_FAILED = (ESP_ERR_ESPNOW_CMP_BASE + 1);
+static const esp_err_t ESP_ERR_ESPNOW_OWN_PEER_ADDRESS = (ESP_ERR_ESPNOW_CMP_BASE + 2);
+static const esp_err_t ESP_ERR_ESPNOW_PAYLOAD_SIZE = (ESP_ERR_ESPNOW_CMP_BASE + 3);
+static const esp_err_t ESP_ERR_ESPNOW_PEER_NOT_SET = (ESP_ERR_ESPNOW_CMP_BASE + 4);
+
+static const esp_err_t ESP_MSG_ESPNOW_BROADCAST = (ESP_ERR_ESPNOW_BASE + 10);
+
+static const esp_err_t CONFIG_ESPNOW_WAKE_WINDOW = 50;
+static const esp_err_t CONFIG_ESPNOW_WAKE_INTERVAL = 100;
+
+ESPNowComponent *ESPNowComponent::static_{nullptr};  // NOLINT
+
+void show_espnow_error(esp_err_t error) {
+  if (error == ESP_ERR_ESPNOW_FAILED) {
+    ESP_LOGE(TAG, "Cannot send espnow packet, espnow is in fail mode.");
+  } else if (error == ESP_ERR_ESPNOW_OWN_PEER_ADDRESS) {
+    ESP_LOGE(TAG, "Trying to send a message to your self.");
+  } else if (error == ESP_ERR_ESPNOW_PAYLOAD_SIZE) {
+    ESP_LOGE(TAG, "Trying to send a payload size to large.");
+  } else if (error == ESP_ERR_ESPNOW_PEER_NOT_SET) {
+    ESP_LOGE(TAG, "Peer address not set.");
+  } else if (error == ESP_ERR_ESPNOW_NOT_INIT) {
+    ESP_LOGE(TAG, "ESPNOW not init.");
+  } else if (error == ESP_ERR_ESPNOW_ARG) {
+    ESP_LOGE(TAG, "Invalid argument");
+  } else if (error == ESP_ERR_ESPNOW_INTERNAL) {
+    ESP_LOGE(TAG, "Internal Error");
+  } else if (error == ESP_ERR_ESPNOW_NO_MEM) {
+    ESP_LOGE(TAG, "Our of memory");
+  } else if (error == ESP_ERR_ESPNOW_NOT_FOUND) {
+    ESP_LOGE(TAG, "Peer not found.");
+  } else if (error == ESP_ERR_ESPNOW_IF) {
+    ESP_LOGE(TAG, "Interface does not match.");
+  }
+}
+
+/* ESPNowPacket ********************************************************************** */
+
+ESPNowPacket::ESPNowPacket(uint64_t peer, const uint8_t *payload, size_t size) {
+  if (size > ESP_NOW_MAX_DATA_LEN) {
+    ESP_LOGE(TAG, "Payload size is to large. It should be less then %d instead it is %d", ESP_NOW_MAX_DATA_LEN, size);
```

Comment:
```suggestion
    ESP_LOGE(TAG, "Payload size %zu is larger than max (%d bytes)", ESP_NOW_MAX_DATA_LEN, size);
```
Shorten log messages but keep the meaning. Also use %zu for size_t. ESP_NOW_MAX_DATA_LEN is plain default defined macro so use %d for it.